### PR TITLE
target/riscv: fix execute_fence

### PR DESCRIPTION
**Pull request by wxjstz**
This patch improves the following issues:
1. Makes it compatible with targets with progbufsize == 1.
2. Although exceptions don’t update any registers, but  do end execution of the progbuf. This will make fence rw, rw impossible to execute.

Change-Id: I2208fd31ec6a7dae6e61c5952f90901568caada6

### CODE REVIEW

**Comment by wxjstz** [comment]
@MarekVCodasip

**Comment by wxjstz** [comment]
[riscv-debug-stable.pdf](https://github.com/riscv/riscv-debug-spec/blob/master/riscv-debug-stable.pdf) page 50 section 4.1
>3. Exceptions don’t update any registers. That includes cause, epc, tval, dpc, and mstatus. They do end execution of the Program Buffer.

**Review by MarekVCodasip — DISMISSED** [review]
Visually, it looks fine.

**Comment by MarekVCodasip** [comment]
Good catch.

**Review by MarekVCodasip — DISMISSED** [review]

**Review by timsifive — CHANGES_REQUESTED** [review]
I think it's worth making two code paths: one which executes both instructions at once (old way), and another which executes them individually (new way).

The reason is that execute_fence() is called from read_memory_progbuf() and write_memory_progbuf(), which can be called quite a lot when a debugger is accessing complex structs from memory.

**Comment by JanMatCodasip** [comment]
> I think it's worth making two code paths: one which executes both instructions at once (old way), and another which executes them individually (new way).

In that case, the algorithm could be:

```
if (progbufsize >= 2 && fence_i_supported):
    execute_fence_i_and_fence_rw_rw()
elif (progbufsize != 0):
    if fence_i_supported:
        execute_fence_i()
    execute_fence_rw_rw()

# The support for fence_i can be detected in examine().
```

@timsifive Is it what you have in mind?

**Comment by wxjstz** [comment]
> I think it's worth making two code paths: one which executes both instructions at once (old way), and another which executes them individually (new way).
> 
> The reason is that execute_fence() is called from read_memory_progbuf() and write_memory_progbuf(), which can be called quite a lot when a debugger is accessing complex structs from memory.

done

**Review by timsifive — APPROVED** [review]

File: src/target/riscv/riscv-013.c [hunk 1]
```diff
@@ -2845,15 +2845,35 @@ static int execute_fence(struct target *target)
 
 	/* FIXME: For non-coherent systems we need to flush the caches right
 	 * here, but there's no ISA-defined way of doing that. */
+	int result;
```
Comment (EccoTheDolphin) [inline src/target/riscv/riscv-013.c:2848]: it looks like we don't really need this `result` variable here. @wxjstz maybe we could cleanup this?

**Pull request merged**